h/admin/views/users.py

```diff
@@ -69,23 +65,59 @@ def users_activate(request):
                                     _query=(('username', user.username),)))
 
 
+@view_config(route_name='admin_users_rename',
+             request_method='POST',
+             permission='admin_users')
+def users_rename(request):
+    user = _form_request_user(request)
+
+    if user is None:
+        return httpexceptions.HTTPFound(
+            location=request.route_path('admin_users'))
+
+    old_username = user.username
+    new_username = request.params.get('new_username')
+
+    try:
+        # FIXME (1) - This assignment validates the username length but not
+        # the character set or any other restrictions.
+        user.username = new_username
+
+        # FIXME (2) - We need to trigger re-indexing of any of the user's
+        # annotations here and ensure that annotation permissions are updated
+        # as a result
+
+        # FIXME (3) - Is there already a service layer that this logic can go
+        # into? If not, should we create one and move the user
+        # rename/activate/delete logic there now?
```

Comment:
Oh, and since Elasticsearch isn't transactional it would be best to change the SQL models, then commit the transaction, and then call the batch indexer. In this case we wouldn't reindex anything when the SQL part fails for whatever reason and minimises the chance of ending up in a bad state.